sips/sip-clarity4/sip-clarity4.md

```diff
@@ -0,0 +1,378 @@
+   back.
+3. **A new Clarity function to convert simple values into `string-ascii`
+   values.** This function will enable developers to easily convert values like
+   `bool`s and `principal`s into their ASCII string representations,
+   facilitating the generation of string-based messages for interacting with
+   cross-chain protocols.
+4. **A new Clarity keyword to get the timestamp of the _current_ block.** This
+   keyword will allow developers to easily access the timestamp of the block
+   currently being processed, enabling time-based logic and features in their
+   smart contracts. This is especially important for DeFi applications.
+
+# Specification
+
+## Fetching the hash of a contract body: `contract-hash?`
+
+Originally proposed [here](https://github.com/clarity-lang/reference/issues/88).
+
+`contract-hash?` returns, on success, the SHA-512/256 hash of the code body of
+the contract principal specified as input. This is useful to prove that a
+deployed contract follows a specific template.
+
+- **Input**: `principal`
+- **Output**: `(response (buff 32) uint)`
+- **Signature**: `(contract-hash? contract-principal)`
+- **Description**: Returns the SHA-512/256 hash of the code body of the contract
+  principal specified as input, or an error if the principal is not a contract
+  or the specified contract does not exist. Returns:
+  - `(ok 0x<hash>)`, where `<hash>` is the SHA-512/256 hash of the code body, on
+    success
+  - `(err u1)` if the principal is not a contract principal
+  - `(err u2)` if the specified contract does not exist
+- **Example**:
+  ```clarity
+  (contract-hash? 'SP2QEZ06AGJ3RKJPBV14SY1V5BBFNAW33D96YPGZF.BNS-V2) ;; Returns (ok 0x9f8104ff869aba1205cd5e15f6404dd05675f4c3fe0817c623c425588d981c2f)
+  ```
+
+## Limiting asset access: `restrict-assets?`
+
+Originally proposed [here](https://github.com/clarity-lang/reference/issues/64).
+
+`restrict-assets?` establishes a context with a deny-all outflow policy for a
+specific principal during the evaluation of its body expressions. It accepts a
+set of allowances, defined using `with-stx`, `with-ft`, `with-nft`, and
+`with-stacking`, which selectively grant outflow allowances. After the body
+expressions finish, the Clarity VM checks the **gross** outflow from the scoped
+principal against the granted allowances. If any allowance is exceeded,
+`restrict-assets?` returns an error; otherwise, it returns `ok` with the result
+of the last body expression.
+
+The most important use case of these new builtins is to allow a contract to
+protect its own assets. To ensure that the contract's assets are safe by
+default, the new `as-contract?` expression is similar to the existing
+`as-contract` expression, but will implicitly behave as though there is a
+`restrict-assets?` expression around its body, specifying the contract principal
+as the asset owner. The old `as-contract` is no longer available in Clarity 4.
```

Comment:
This is interesting  - i think this should be added to the backwards compatibility section.

Reply:
Added in 573917b50ed47cdb49ba2887c765c97918b16beb.